planner/core/rule_column_pruning.go

```diff
@@ -361,6 +360,15 @@ func (la *LogicalApply) PruneColumns(parentUsedCols []*expression.Column) error
 	}
 
 	la.mergeSchema()
+
+	if la.JoinType == LeftOuterSemiJoin || la.JoinType == AntiLeftOuterSemiJoin {
+		joinCol := la.schema.Columns[len(la.schema.Columns)-1]
+		parentUsedCols = append(parentUsedCols, joinCol)
+	} else if la.JoinType == LeftOuterJoin {
+		parentUsedCols = append(parentUsedCols, la.children[0].Schema().Columns...)
```

Comment:
In my opinion, for correlated subquery, the main logical plan optimizes is decorrelation. The core idea is: push the Apply operator down, and pull up the operator(innerPlan operator) under Apply as much as possible. That means the father node of the Apply operator may change. 
Now we will inline the projection for Apply, which means the father of Apply operator is a Projection operator, it may come from the origin plan, or from the result of decorrelation. If the Projection operator comes from the latter, the schema of the Projection will set by the Apply operator, see: 
https://github.com/pingcap/tidb/blob/cc0bef82b72b6ce6b3138e76f55db5f5eb5a5d7d/planner/core/rule_decorrelate.go#L156
For `LeftOutJoin`, we need to keep all left child's columns in the result, so it should be added in the `parentUsedCols` to make sure the columns not be pruned by the `la.inlineProjection(parentUsedCols)`.

Reply:
Optimizer inlines projection twice, and the first one runs before `rule_decorrelate`. If any columns of innerPlan is eliminated, it seems that schema & `Exprs` of projection may be inconsistent, and cause error.

So I think we need 1) Do inline projection for `Apply` in second inline-projection only, which runs after de-correlation.
Or 2) Aware that some columns in schema of `Apply` may have been eliminated, and the elimination should be done to `Exprs` of projection too.